[Swift in WebKit] Add a modulemap for the private interface of JSC

rr-codes opened the pull request:
#### 0d9b3a0d2ce0834de493974c9e5cffd394bcdc10
<pre>
[Swift in WebKit] Add a modulemap for the private interface of JSC
<a href="https://bugs.webkit.org/show_bug.cgi?id=298124">https://bugs.webkit.org/show_bug.cgi?id=298124</a>
<a href="https://rdar.apple.com/159472949">rdar://159472949</a>

Reviewed by Keith Miller.

Add a private module map to JSC, a prerequisite for modularizing WebCore.

* Source/JavaScriptCore/API/JSContextInternal.h:
* Source/JavaScriptCore/API/JSContextPrivate.h:
* Source/JavaScriptCore/API/JSScript.h:
* Source/JavaScriptCore/API/JSValueInternal.h:
* Source/JavaScriptCore/API/JSValuePrivate.h:
* Source/JavaScriptCore/API/JSVirtualMachinePrivate.h:

Add some missing OBJC guards, so that JSC can be compiled in C++ only mode.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:

- Ensure the module verifier runs on the public SDK to make it easier to catch issues
- Adjust the module verification flags to use the built WTF/bmalloc modules instead of the system ones
- Disable verification for Obj-C and C. This is done because the JSC private interface would require
the addition of __cplusplus guards to every private header file. This itself wouldn&apos;t be needed usually,
however the module verifier does not support module require clauses; if this functionality becomes supported,
or if a custom verifier is written, these languages can be re-added. It is reasonably safe to omit this
subset of languages since ObjC++ and C++ are still covered, and it is unlikely for C++ API to accidentally
be added to the public JSC API.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/JavaScriptCore_Private.modulemap: Added.

Canonical link: <a href="https://commits.webkit.org/300796@main">https://commits.webkit.org/300796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3541eaf5122515259bd37410f85f101d374d1d28

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123911 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43626 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34323 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130698 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b8912c1c-c07b-4b4c-9065-c5ac287d4a46) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44350 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52220 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c4d09a70-2f40-4559-89a9-585b22ff9bd9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126865 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35332 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74846 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2d1b0d68-2ea2-44b3-be1c-cb22b0358bbb) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/29004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74174 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116070 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105060 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/29227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133393 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122443 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50864 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51238 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/107063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26067 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47883 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50717 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56481 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153539 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39021 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/53537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/JavaScriptCore/JavaScriptCore_Private.modulemap:
```diff
@@ -0,0 +1,16 @@
+//  JavaScriptCore_Private.modulemap
+//  JavaScriptCore
+//
+//  Created by Richard Robinson on 8/29/25.
```
fix this

kmiller68 reviewed (approved):
r=me when EWS is green

adetaylor commented on Source/JavaScriptCore/JavaScriptCore_Private.modulemap:
```diff
@@ -0,0 +1,93 @@
+        export *
+    }
+
+    explicit module Cxx {
```
FWIW I've started to move away from using the keyword `explicit`. I'm still not sure of the pros and cons, but I've found no advantages to using the keyword, and presumably it might require somebody in future to do something like `import JavaScriptCore.Cxx` which is probably undesirable.

rr-codes committed:
[Swift in WebKit] Add a modulemap for the private interface of JSC
https://bugs.webkit.org/show_bug.cgi?id=298124
rdar://159472949

Reviewed by Keith Miller.

Add a private module map to JSC, a prerequisite for modularizing WebCore.

* Source/JavaScriptCore/API/JSContextInternal.h:
* Source/JavaScriptCore/API/JSContextPrivate.h:
* Source/JavaScriptCore/API/JSScript.h:
* Source/JavaScriptCore/API/JSValueInternal.h:
* Source/JavaScriptCore/API/JSValuePrivate.h:
* Source/JavaScriptCore/API/JSVirtualMachinePrivate.h:

Add some missing OBJC guards, so that JSC can be compiled in C++ only mode.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:

- Ensure the module verifier runs on the public SDK to make it easier to catch issues
- Adjust the module verification flags to use the built WTF/bmalloc modules instead of the system ones
- Disable verification for Obj-C and C. This is done because the JSC private interface would require
the addition of __cplusplus guards to every private header file. This itself wouldn't be needed usually,
however the module verifier does not support module require clauses; if this functionality becomes supported,
or if a custom verifier is written, these languages can be re-added. It is reasonably safe to omit this
subset of languages since ObjC++ and C++ are still covered, and it is unlikely for C++ API to accidentally
be added to the public JSC API.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/JavaScriptCore_Private.modulemap: Added.

Canonical link: https://commits.webkit.org/300796@main
